crates/cow-amm/src/indexer.rs

```diff
@@ -0,0 +1,162 @@
+    events_registry: Arc<RwLock<EventsRegistry>>,
+    pub first_block: u64,
+}
+
+impl Indexer {
+    pub async fn new(
+        web3: &DynWeb3,
+        cow_amm_factory_contract: Option<&cow_amm_constant_product_factory::Contract>,
+    ) -> Self {
+        let cow_amm_constant_product_factory =
+            if let Some(cow_amm_factory_contract) = cow_amm_factory_contract {
+                cow_amm_factory_contract.clone()
+            } else {
+                contracts::CowAmmConstantProductFactory::deployed(web3)
+                    .await
+                    .expect("Failed to find deployed CowAmmConstantProductFactory")
+            };
+        let first_block = Self::first_block(cow_amm_constant_product_factory)
+            .await
+            .unwrap_or(0);
+        Self {
+            cow_amms: Arc::new(RwLock::new(HashMap::new())),
+            events_registry: Arc::new(RwLock::new(BTreeMap::new())),
+            first_block,
+        }
+    }
+
+    async fn first_block(
+        cow_amm_constant_product_factory: cow_amm_constant_product_factory::Contract,
+    ) -> Option<u64> {
+        let first_block = match cow_amm_constant_product_factory.deployment_information()? {
+            DeploymentInformation::BlockNumber(block) => block,
+            _ => return None,
+        };
+        Some(first_block)
+    }
+
+    /// Returns all CoW AMMs that are currently enabled (i.e. able to trade).
+    pub async fn cow_amms(&self) -> Vec<impl crate::CowAmm> {
+        let cow_amms = self.cow_amms.read().await;
+        cow_amms.values().cloned().collect::<Vec<_>>()
+    }
+}
+
+#[async_trait::async_trait]
+impl EventStoring<cow_amm_constant_product_factory::Event> for Indexer {
+    async fn replace_events(
+        &mut self,
+        events: Vec<ethcontract::Event<cow_amm_constant_product_factory::Event>>,
+        range: RangeInclusive<u64>,
+    ) -> anyhow::Result<()> {
+        // Context to drop the write lock before calling `reapply_events()`
+        {
+            let range = *range.start()..=*range.end();
+            let events_registry = &mut *self.events_registry.write().await;
+
+            // Remove the CowAmmConstantProductFactoryEvent events in the given range
+            for key in range {
+                events_registry.remove(&key);
+            }
+        }
+
+        // Revert the CoW AMM status after removing the events in the specified
+        // range. Since we cannot know what is the CoW AMM status before an
+        // event is reverted, in order to guarantee the CoW AMM status is
+        // correct, we need to reapply all the events and substitute
+        // the current CoW AMM in memory after the reapplication.
+        // It returns the new list of CoW AMMs after all the events in the
+        // registry have been applied
+        // Context to drop the write lock before calling `reapply_events()`
+        {
+            let events_registry = &mut *self.events_registry.write().await;
```

Comment:
This is the same write lock you already took in the scope above. Also this could be skipped with the type simplification proposed above.